### PR TITLE
Added a fix for broken chaincode dependencies

### PR DESCRIPTION
**Pull request by masterDev1985**
During the process of collecting the chaincode specified in the "path" parameter of the POST request.  A temporary directory is created to store the chaincode in before it is built.  This chaincode depends on obc-peer code (specifically the shim code) in order to work, which triggers a pull of the obc-peer repo.  Because the repository is private, the pull request will query the command line for user credentials, which will of course never be entered, causing the deployment process to go off into the ether, as far as the user deploying code is concerned.

All of this can be avoided by copying the local copy of obc-peer into the temporary directory before the go get command.  This is the change that I have implemented.  This is obviously not an optimal solution, but it's a good workaround.

Signed-off-by: Dale Avery davery@us.ibm.com

@muralisrini

### CODE REVIEW

**Comment by masterDev1985** [comment]
Signed-off-by: Dale Avery davery@us.ibm.com

**Comment by muralisrini** [comment]
@masterDev1985 : while the workaround looks like a good idea, I hesitate to put it out - even as a workaround .... we are violating user intent by using a different obc code base. Worse, the original "go get" command is broken as implemented. Let us get together and figure out the correct way to fix this.

**Comment by masterDev1985** [comment]
@muralisrini sounds good.  Thanks for the feedback.  I'm not sure I understand what you mean by the go get command being broken.  I'll close this pull request, but I do want to stress that the REST API does not currently work without this fix.

File: core/chaincode/platforms/golang/list.go [hunk 1]
```diff
@@ -68,16 +70,15 @@ func gopathDependencyPackageInfo(goos, goarch, pkg string) ([]PackageInfo, error
...
+		return nil, errors.Wrapf(err, "listing deps for package %s failed", pkg)
 	}
-	decoder := json.NewDecoder(stdout)
-
-	err = cmd.Start()
-	if err != nil {
-		return nil, err
+	// on Windows make sure we use forward slashes
+	if os.PathSeparator == '\\' {
+		output = []byte(filepath.ToSlash(string(output)))
```
Comment (sykesm) [inline core/chaincode/platforms/golang/list.go:79]: Modifying the json before unmarshaling isn't a safe operation. It's being fed far more than a single path.

File: core/chaincode/platforms/golang/list.go [hunk 2]
```diff
@@ -68,16 +70,15 @@ func gopathDependencyPackageInfo(goos, goarch, pkg string) ([]PackageInfo, error
 	cmd := exec.CommandContext(ctx, "go", "list", "-deps", "-f", packageListFormat, pkg)
 	cmd.Env = append(os.Environ(), "GOOS="+goos, "GOARCH="+goarch)
 
-	stdout, err := cmd.StdoutPipe()
+	output, err := cmd.Output()
```
Comment (sykesm) [inline core/chaincode/platforms/golang/list.go:73]: We're increasing memory load for little reason by reading all output up front.  A decoder was originally used to process the stream as it was written to reduce memory overhead (streaming) and to support multiple objects in the list output.

File: core/chaincode/platforms/golang/list.go [hunk 3]
```diff
@@ -7,12 +7,14 @@ SPDX-License-Identifier: Apache-2.0
 package golang
 
 import (
+	"bytes"
```
Comment (sykesm) [inline core/chaincode/platforms/golang/list.go:10]: I don't understand why any changes are needed in list. We are just capturing the output from the tool to generate a model.

I do agree we need to make changes in how we deal with the model.

File: core/chaincode/platforms/golang/list.go [hunk 4]
```diff
@@ -68,16 +70,15 @@ func gopathDependencyPackageInfo(goos, goarch, pkg string) ([]PackageInfo, error
...
-		return nil, wrapExitErr(err, "'go list -deps' failed")
+		return nil, errors.Wrapf(err, "listing deps for package %s failed", pkg)
 	}
-	decoder := json.NewDecoder(stdout)
-
-	err = cmd.Start()
-	if err != nil {
-		return nil, err
+	// on Windows make sure we use forward slashes
+	if os.PathSeparator == '\\' {
```
Comment (sykesm) [inline core/chaincode/platforms/golang/list.go:78]: If we do need to make changes here, no need to check path separators; just use ToSlash - it's idempotent.

File: core/chaincode/platforms/golang/list_test.go [hunk 5]
```diff
@@ -9,6 +9,7 @@ package golang
 import (
```
Comment (sykesm) [inline core/chaincode/platforms/golang/list_test.go:9]: Assuming there are no changes required in list.go, there shouldn't' be any changes required here either (beyond fixing the spelling).

**Pull request closed without merging**